arxiv-opensearch-cohere/README.md

```diff
@@ -0,0 +1,256 @@
+    text = row.abstract
+    try:
+        embed = cache[text]
+        body = {
+            VECTOR_NAME: embed,
+            "text": text,
+            "title": row.title,
+            "arxiv_id": row.id,
+            "doi": row.doi,
+        }
+        response = client.index(index=INDEX_NAME, id=i, body=body)
+    except Exception as e:
+        print(f"[ERROR]: {e}")
+        continue
+```
+We have included the `abstract`, `title`, `arxiv_id` and `doi` fields in the document index from the data file. Feel free to include more or less depending on what you think you would want to search for. 
+
+Using the [opensearch-py-ml](https://pypi.org/project/opensearch-py-ml/) package, we are able to query our OpenSearch cluster with a pandas-like interface. You can run the following to ensure the documents inserted at your `index_name` are what you expect in terms of size. 
+
+**Note:** when using `opensearch-py-ml` package, ensure you look through the [documentation](https://opensearch-project.github.io/opensearch-py-ml/reference/index.html) as you cannot do every command as you would in pandas. 
+
+```python
+oml_df = oml.DataFrame(client, INDEX_NAME)
+print(oml_df.shape)
+```
+
+Your index has been created! 
+
+Full script can be run with `python create_index.py`. 
+
+## Step 4: Query your index for similar documents using cohere embeddings
+
+Now that your index is populated, you can proceed to querying it!
+Since we are using vectors for the kNN search, any of our queries must be translated to a vector using Cohere. The query vector is then submitted to the query endpoint within OpenSearch to find similar vectors via Approximate k-NN. 
+
+We can do so with the following helper functions: 
+```python
+
+def get_cohere_embedding(text: str) -> List[float]:
+    """
+    Embed a single text with cohere client and return list of float32
+    """
+    embed = np.array(co.embed([text], model=COHERE_MODEL).embeddings)
+    return list(embed.reshape(-1))
+
+
+def find_similar_docs(query: str, k: int, num_results: int, index_name: str) -> Dict:
+    """
+    Main vector search capability using knn on input query strings.
+    Args:
+        k: number of k-similar neighbors/vectors to retrieve from opensearch index
+        num_results: number of the k-similar vectors to retrieve.
+        index_name: index name in opensearch
+    """
+    embed_vector = get_cohere_embedding(query)
+
+    body = {
+        "size": num_results,
+        "query": {"knn": {VECTOR_NAME: {"vector": embed_vector, "k": k}}},
+    }
+
+    url = f"http://localhost:9200/{index_name}/_search"
+    response = requests.get(
+        url, json=body, headers={"Content-Type": "application/json"}
+    )
+    return json.loads(response.content)
+```
+
+The above functions search our index using 2 important parameters: 
+* `k` = the number of neighbors the `hnsw` search will return per query. The maximum `k` supported is 10,000. 
+* `size` = how many results will be returned from the query
```

Comment:
what's the difference between k and size?  it looks at k NNs and to find the closest oines and then returns size of the closest?

Reply:
my understand is `k` is how many neighbors you want ie, for every query, I want 2 closest neighbors 
and `size` is how many of those top-2 do you want, ie, I want 10 results of the top-2 neighbors. 
Depending on the corpus, they could all be the same result or a variety

Reply:
not following. we probably need to discuss.